cpu/esp8266: esp-now network device support

gschorcht opened the pull request:
### Contribution description

This PR introduces the ESP-NOW netdev device driver. The ESP-NOW netdev device allows ESP8266 nodes module to communicate using the built-in WiFi interface and the GNRC stack.

### Test procedure

Since ESP-NOW is a technology that only works between ESP nodes, at least two esp8266 boards are needed for testing. The easiest way to test is to use ```examples/gnrc_networking```. Once the boards have been flashed with the following make command
```
USEMODULE=esp_now make flash -C examples/gnrc_networking BOARD=...
```
the commands ```ifconfig``` and ```ping6``` can be used for testing.

### Issues/PRs references

This PR depends on ~#10126~, ~#10536~, ~#10581~, ~#10656~,  ~#10700~.

smlng commented on cpu/esp8266/include/periph_cpu.h:
```diff
@@ -160,13 +160,13 @@ typedef enum {
  */
 #define PERIPH_TIMER_PROVIDES_SET
 
-/** @} */
```
good catch, THX!

emmanuelsearch commented:
@gschorcht nice! How does one test this PR?

gschorcht commented:
@emmanuelsearch Please refer the updated description.

gschorcht commented:
@A-Paul Many thanks for your test.

Don't wonder please, when I was investigating your problems, I realized some minor inconsistencies in some functions when debugging output is enabled. So, I have fixed them with the last commit and would like to include them with this PR. These are in detail:

- ```hexdump``` was renamed somewhere in the past to ```esp_hexdump``` (```cpu/esp8266/esp-now/esp_now_netdev```)
- missing spaces after format specifier in ```esp_hexdump```  (```cpu/esp8266/tools.c```)
- hex format specifier are not supported by ```ets_printf``` (```cpu/esp8266/include/sys/_intsup.h```)
- module ```netopt``` is required when debug output is enabled (```cpu/esp8266/Makefile.include```)

As stated, I could realize these inconsistencies only after setting ENABLE_DEBUG.

miri64 commented:
> That behavior seems to be caused by upper layers of GNRC or the application. esp-now netdev allows messages with a maximum length of 250 bytes and. Since it is of type NETDEV_TYPE_RAW, IPv6 packets are embedded without changes in esp-now messages. Since the IPv6 header has 40 byte and the UDP header has 8 byte, you can send 202 bytes at maximum as UDP payload. The maximum packet size is reported correctly by the esp-now netdev with NETOPT_MAX_PACKET_SIZE and seems to be is used by upper layers. Although you see the message Success: send 203 byte to ..., the _send function of the esp-now netdev is not called. Probably, the size is checked in the upper layer. Maybe, @miri64 can explain why you a success message instead of an an error message when overlength packet are tried to be sent. Maybe the return code is not handled in tests/gnrc_udp/main.c.

Due to the asynchronous nature of GNRC there is no error reporting (basically sending is just fire-and-forget). There is an experimental `gnrc_neterr` module that helps reporting errors, but don't think there is any application supporting it yet.

gschorcht commented:
@A-Paul How do we proceed, the test was successful and the observed behavior isn't caused by the driver. Would you give your OK to get it built with murdock?

A-Paul commented:
Hi @gschorcht,
Because I have a bunch of nodemcu clones lying around, I wanted to provide a quick "hands on" report in the first place. At this moment I don't have the time to look deeper into your code and hence don't have enough insight to approve the PR alone.

I might repeat the quick test based on your last changes during this weekend.

@smlng, i seems you had a look into the code already. What is your affirmation level? :)

A-Paul commented:
Repeated tests as in https://github.com/RIOT-OS/RIOT/pull/9917#pullrequestreview-159098311
Connected three *esp8266-esp-12x*. *ping6*, *udp server/send* or *nib neigh show* look plausible.

jia200x commented:
any news here? Feature Freeze is tomorrow

PeterKietzmann commented:
So far this PR did not get a proper code review and neither in-depth testing. I propose we skip this feature for the release but try to get it merged somewhat soon so we have a stable version in the next release. I suppose **many** people are interested in networking with ESPs so let's not take the risk of merging prematurely.

TimoRoth commented:
I wonder, could the esp-now driver from esp32, which looks largely identical to this one, be moved to a common location, so the driver code isn't duplicated for the two CPUs? Or is there something different between the two that I'm missing?

gschorcht commented:
Changes as in PR #10581 required.

miri64 commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -869,6 +869,18 @@ int gnrc_netif_ipv6_get_iid(gnrc_netif_t *netif, eui64_t *eui64)
+                eui64->uint8[5] = netif->l2addr[3];
+                eui64->uint8[6] = netif->l2addr[4];
+                eui64->uint8[7] = netif->l2addr[5];
+                return 0;
```
I believe this isn't needed anymore ;-)

miri64 commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-6ln.c:
```diff
@@ -69,6 +69,16 @@ static inline uint8_t _reverse_iid(const ipv6_addr_t *dst,
+            l2addr[4] = dst->u8[14];
+            l2addr[5] = dst->u8[15];
+            return ETHERNET_ADDR_LEN;
+#endif  /* MODULE_ESP_NOW */
```
Dito ;-)

miri64 commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-arsm.c:
```diff
@@ -204,6 +204,11 @@ static inline unsigned _get_l2addr_len(gnrc_netif_t *netif,
+        case NETDEV_TYPE_RAW:
+            (void)opt;
+            return ETHERNET_ADDR_LEN;
+#endif  /* MODULE_ESP_NOW */
```
Dito

miri64 commented:
Sorry, only realized now, that you set this PR back to WIP :sweat_smile:

gschorcht commented:
@miri64 Sorry, only realized now, that you set this PR back to WIP

Yes, this PR requires the same changes as ESP-NOW for ESP32 :worried: I will do that during the holiday season. I hope I can reuse a lot.

gschorcht commented:
The PR is now a rework that reuses @TimoRoth's changes of the ESP32 esp_now netdev driver as provided in PR #10581.

@TimoRoth 
> I wonder, could the esp-now driver from esp32, which looks largely identical to this one, be moved to a common location, so the driver code isn't duplicated for the two CPUs? Or is there something different between the two that I'm missing?

Indeed, most parts are the same for ESP32 and ESP8266. While there are only slight differences in most functions, the initialization in function `netdev_esp_now_setup` is completely different. We could try to introduce `cpu/common/esp` but I'm not sure whether this is worth for the moment for this only one module. We should probably keep it separate for the moment, so that it can be merged into the master for version 2019.01.

miri64 commented:
> We should probably keep it separate for the moment, so that it can be merged into the master for version 2019.01.

If just one function differs, I think it is cleaner to wait one release cycle and put the extra effort into merging the two implementations and putting `netdev_esp_now_setup` in a separate module for each CPU type.

gschorcht commented:
@miri64 Hm, it's a bit more than just one function that's different, see [esp_now_netdev.c.diff](https://gist.github.com/gschorcht/4caaa82188ee39bf181514c9e7bbbb80). The problem is that the ESP-NOW APIs are very similar but have following differences:

- Available functions and how they are used are similar but slightly different.
- ESP8266 version of the API does not define const parameters while the ESP32 version does
- Configuration and resulting data structures are different.

To use modules together for both MCUs, some changes in makefiles are required. I have this already on my ToDo list, issue #10658. But for the moment, I would be happy if we could leave it as it is.

ESP-NOW is the only possible built-in networking interface on ESP8266. Together with an ESP32 border router, ESP8266 nodes would get connectivity. Therefore, IMHO it is important to get it into the release.

Please take a quick look at the diff file and decide whether it is worth to merge these two versions into a shared module. If so, I can try to change the makefiles in the short term.

miri64 commented:
Sorry, I somehow was under the impression that this re-implements the whole `gnrc_netif` sub-layer...

miri64 commented:
Ah it *does*. Your diff just didn't show it because they are the same... @TimoRoth @gschorcht Can't those be moved to `sys/net/gnrc/netif/esp-now/` at least (as a preparing step for both this PR and #10581), so we avoid that huge chunk of code duplication?

gschorcht commented:
Yes, `esp_now_netif` is exactly the same. `esp_now_netdev` differs.

> Can't those be moved to `sys/net/gnrc/netif/esp-now/ at least`

Sure, but I'm not sure whether it is a good idea. The `netif` implementation there are the standard techologies, Ethernet and IEEE802.15.4. ESP-NOW is a proprietary technologie only used for ESP nodes.

Let me try to change the makefiles, to see how much effort it is to make it possible that ESP32 and ESP8266 can use shared modules in `cpu/common/esp`.

miri64 commented:
> Sure, but I'm not sure whether it is a good idea. The `netif` implementations there are standard techologies, Ethernet and IEEE802.15.4. ESP-NOW is a proprietary technologie only used for ESP nodes.
> 
> Let me try to change the makefiles, to see how much effort it is to make it possible that ESP32 and ESP8266 can use shared modules in `cpu/common/esp`.

I'm fine with either. I just don't want to merge 250+ lines of code duplication, but common code should go into common places. Otherwise we just have to cleanup stuff later and from experience that takes oftentimes longer than actually doing it right from the start ;-).

miri64 commented on cpu/esp8266/esp-now/esp_now_netif.c:
```diff
@@ -0,0 +1,251 @@
+/*
```
Also maybe use someting like `esp_now_gnrc`, because it isn't the netif part that is of interest here (the implementation below wouldn't work with `lwip_netif` e.g.).

gschorcht replied:
This should be changed in #10581 first.

gschorcht commented:
@miri64 @TimoRoth I provided a PR #10663 with the makefile changes and moved the old version of esp_now netdev driver from `cpu/esp32` to `cpu/esp_common` to demonstrate on how to use it.

@miri64 Do you think we can get it merged in short term to be able to continue with the other PRs? Succesful compilation by murdock should be a reasonable test for it.

@TimoRoth Would you change your PR to the new location or should I do it by providing a PR to your fork?

TimoRoth commented:
I'll look into it tomorrow when I get home from family stuff again.

gschorcht commented:
@miri64 @TimoRoth I merged the ESP8266 esp_now netdev implementation with `cpu/esp_common/esp-now`. You might want to look at it.

gschorcht commented:
The main part is commit https://github.com/RIOT-OS/RIOT/pull/9917/commits/f412c9101d5158014f88f58559abe710f59da64c.

TimoRoth commented:
Code-Wise it looks very reasonable on first quick glance. I'll give it a try later or tomorrow.

miri64 commented:
Oops, sorry for the commit comment :-/

miri64 commented on cpu/esp_common/esp-now/esp_now_netdev.c:
```diff
@@ -53,21 +61,48 @@
+#define IRAM_ATTR                IRAM
+#endif
+
+#define esp_const_uint8_t        uint8_t
```
Let's move [this discussion](https://github.com/RIOT-OS/RIOT/commit/f412c9101d5158014f88f58559abe710f59da64c#diff-24495fdc92b602d8c37d024123b775a5R83) back to the PR ;-)

miri64 replied:
I'd prefer not to hide the non-constness. If it is required to be not const for certain implementations, then remove the const qualifier completely.

gschorcht replied:
Then, the only way is to embed functions with different signatures in `#ifdef MCU_ESP32 .. #else ... #endif`

miri64 replied:
> > Not const :-/
> 
> Yes, that's the trick to be able to use as much as possible shared code and not to need to embed all function signatures in
> 
> ```c
> #ifdef MCU_ESP32
> static IRAM_ATTR void esp_now_recv_cb(const uint8_t *mac, const uint8_t *data, int len)
> #else
> static IRAM_ATTR void esp_now_recv_cb(uint8_t *mac, uint8_t *data, uint8_t len)
> #endif
> ```
> 
> Unfortunately, even the ESP-NOW API function seems to be the same, the parameters of ESP8266 version are always non-const while the parameters of ESP32 are const 
> 
> This is especially a problem when a certain function signature is required when they are registered as callbacks.

> Then, the only way is to embed functions with different signatures in `#ifdef MCU_ESP32 .. #else ... #endif`

I'd prefer this over intransparent type magic. Alternative: move these functions to CPU-specific implementation?

gschorcht replied:
The problem is only the signature, implementations are at least partially the the same.

gschorcht replied:
Changed.

miri64 replied:
I'll have another look in a moment

gschorcht commented:
I have rebased (and merged) on to the current master which contains the fixes and optimization changes #10700 made for ESP32. It was tested again with ESP8266 as well ESP32. From my point of view, it is ready to be merged and would have to be squashed.

Unfortunatly, #10656 is a prerequisite, but #10656 is also rebased on to current master and ready to be merged.

gschorcht commented:
@smlng What about this PR? PR #10792 does not depend on it but it is related.

smlng commented on cpu/esp8266/Makefile.include:
```diff
@@ -24,6 +24,22 @@ endif
+    USEMODULE += esp_sdk
+endif
+
+ifneq (, $(filter esp_now, $(USEMODULE)))
```
this is also in #10792, so we should wait and merge that first

gschorcht commented:
Rebased. It is still working. Once this PR is merge, I will provide a PR with a cleanup which is needed to get `esp_wifi` and `esp_now` working together for border routers.

gschorcht commented:
How do we proceed with this PR? All requests for changes were considered some time ago.

ESP-NOW for ESP8266 is compatible with ESP-NOW for ESP32 and is a prerequisite to use ESP32 and ESP8266 in the same network. ESP-NOW allows that ESP nodes to communicate without having infrastructure mode WiFi. And finally, this PR is prerequisite for the cleanup in PR #10929 and would make #10980 obsolete.

tcschmidt commented:
@miri64 @smlng - comments addressed? How about practical testing?

smlng commented:
will test today, need to find a 2nd esp8266 though.

gschorcht commented:
@smlng Thanks :smile: BTW, testing should also be possible with one ESP8266 and one ESP32 node.

smlng commented on cpu/esp8266/include/common.h:
```diff
@@ -83,6 +83,12 @@ extern "C" {
 #endif
 
+
+#define LOG_TAG_ERROR(tag, fmt, ...)   LOG_ERROR("[%s] " fmt, tag, ##__VA_ARGS__)
```
unrelated? These are never used by `esp8266` code (yet).

gschorcht replied:
Not used in `cpu/esp8266` but in `cpu/esp_common/esp_now` netdev, e.g., here https://github.com/RIOT-OS/RIOT/blob/ac775b8ae5e3f5a385239a16ebbdcb505b1cae89/cpu/esp_common/esp-now/esp_now_netdev.c#L351

That is, when `esp_now` is used, they are required. These macros will be moved later to `cpu/esp_common` in further cleanup PRs. But for the moment and this PR, they are required in `cpu/esp8266`.

smlng commented:
@miri64 good to get this merged?

miri64 commented:
@gschorcht I think you can squash ;-)

gschorcht commented:
squashed

gschorcht commented:
@smlng Is it ok to merge?

gschorcht commented:
Thanks.
